### PR TITLE
Add --replace flag to podman secret create

### PR DESCRIPTION
**Pull request by rhatdan**
Users may want to update the secret used within containers, without destroying the secret and recreating it.

Partial fix for https://github.com/containers/podman/issues/18667

#### Does this PR introduce a user-facing change?

```release-note
Podman secret create now has a --replace option, which allows you to modify secrets without replacing containers.
```

### CODE REVIEW

**Comment by ashley-cui** [comment]
I feel like update is a different enough that it should be a separate secret command? I think the way update in common is written is that the secret ID will not change on an update, and `podman secret create --replace` implies that the secret is a new secret, with a new ID, but it replaces the old secret with the name.

This is more of a design decision though, I'm happy to hear other opinions

**Comment by rhatdan** [comment]
I went with replace since we use this one for `podman container create --replace`

**Comment by ashley-cui** [comment]
This would break containers that use the secret. 

I believe container create --replace does change the container ID.

**Comment by ashley-cui** [comment]
Well, kind of breaks. If I recall correctly the once a container is created with a secret, the secret is never updated in the container, even if the secret is deleted or changed. It would only break if a user re-ran the container when specifying a secret using its ID.

**Comment by rhatdan** [comment]
So you are saying the secret is embedded in the container, restarting the container will not get the new secret?

**Comment by ashley-cui** [comment]
Yes, that's correct. I believe that's why we didn't implement updating secrets in the first pass of secrets because it wouldn't change the secret inside created containers in the first place.

**Comment by rhatdan** [comment]
@kriansa is this what you intended? IE Change the password in existing containers or are you thinking just for newly created containers?

**Comment by kriansa** [comment]
I'm fine with how it's been implemented given the complexity of only doing it in newly created containers. In many cases, applications are not prepared to hot reload secrets anyway and it's common to redeploy (i.e. recreate) them just to ensure they load up the correct values on initialization. See for example [reloader], which does exactly that on Kubernetes.

For those relying on stateful containers, we might want to show a warning message when you attempt to replace a secret currently in use. WDYT?

[reloader]: https://github.com/stakater/Reloader

File: pkg/bindings/secrets/types_create_options.go
Comment (BlackHole1) [inline pkg/bindings/secrets/types_create_options.go:90]:
```suggestion
		return false
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: pkg/bindings/secrets/types_create_options.go
Comment (rhatdan) [reply pkg/bindings/secrets/types_create_options.go:90]: These are generated, so if you change them, you need to hcnage them else where.

Comment (BlackHole1) [reply pkg/bindings/secrets/types_create_options.go:90]: Okay.

**Comment by vrothberg** [comment]
> For those relying on stateful containers, we might want to show a warning message when you attempt to replace a secret currently in use. WDYT?

Great question!  Currently, `podman secret rm` does not check whether a secret's in use, so I think `--replace` should just follow `rm`.

File: docs/source/markdown/podman-secret-create.1.md
Comment (vrothberg) [inline docs/source/markdown/podman-secret-create.1.md:45]: I think we need to be more verbose. Compare to `container create --replace`:

> If another container with the same name already exists, replace and remove it. The default is false.

Comment (vrothberg) [reply docs/source/markdown/podman-secret-create.1.md:45]: same for the --help message

**Review by vrothberg — COMMENTED** [review]
Other than the doc nits, LGTM

**Comment by ashley-cui** [comment]
/lgtm
/hold

Thanks! Sorry for the fuss on naming the flag, I just want to make sure we're consistent in our flags between our commands

File: pkg/api/handlers/libpod/secrets.go
Comment (ashley-cui) [inline pkg/api/handlers/libpod/secrets.go:27]:
```suggestion
		Update    bool              `schema:"update"`
```

File: pkg/api/handlers/libpod/secrets.go
Comment (ashley-cui) [inline pkg/api/handlers/libpod/secrets.go:40]:
```suggestion
	opts.Update = query.Update
```

**Review by ashley-cui — COMMENTED** [review]
Oops, didn't take close enough a look, but once this is fixed I'll slash lgtm it if needed

File: docs/source/markdown/podman-secret-create.1.md
Comment (vrothberg) [inline docs/source/markdown/podman-secret-create.1.md:47]:
```suggestion
The default is **false**.
```

File: test/e2e/secret_test.go
Comment (vrothberg) [inline test/e2e/secret_test.go:40]: Can we also test that labels are updated?  Create the initial one with `label foo` and the updated one with `label bar`.

We should also document whether the labels will be entirely replaced or added.

**Comment by vrothberg** [comment]
> Oops, didn't take close enough a look, but once this is fixed I'll slash lgtm it if needed

I'd love more pairs of eyes on the code given yesterday's discussion around naming etc.

@Luap99 @mheon PTAL

File: pkg/domain/infra/tunnel/secrets.go
Comment (BlackHole1) [inline pkg/domain/infra/tunnel/secrets.go:19]:
```suggestion
		WithUpdate(options.Update)
```

File: test/e2e/secret_test.go
Comment (BlackHole1) [inline test/e2e/secret_test.go:45]:
```suggestion
		session = podmanTest.Podman([]string{"secret", "create", "-d", "file", "--driver-opts", "opt1=val1", "--update", "a", secretFilePath})
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]
The PR title and release note in the PR body still need to be modified.

**Comment by rhatdan** [comment]
We are sticking with --replace, not --update.  If we want to add an update mechanism I think we should add
`podman secret update ` command and then have it affect all containers. (Optionally)

**Comment by vrothberg** [comment]
> We are sticking with --replace, not --update. If we want to add an update mechanism I think we should add `podman secret update ` command and then have it affect all containers. (Optionally)

`--replace` should then create an entirely new secret (and ID) to be consistent with other commands.

**Comment by rhatdan** [comment]
Thats the plan

**Comment by vrothberg** [comment]
Thanks!

**Comment by rhatdan** [comment]
@vrothberg Something strange is going on with containers/common. The older released version keeps getting sucked in unless I do a `go mod edit --replace`

**Comment by TomSweeneyRedHat** [comment]
I think this is the PR fix you made in common to address this: https://github.com/containers/common/pull/1541  If so, it went into the main branch and not the v0.55 branch.  I think you need a cherry-pick there.

**Comment by vrothberg** [comment]
> I think this is the PR fix you made in common to address this: [containers/common#1541](https://github.com/containers/common/pull/1541) If so, it went into the main branch and not the v0.55 branch. I think you need a cherry-pick there.

This is for a new feature that is not part of Podman 4.6.  I don't think it needs backporting to the v0.55 branch.

**Pull request merged**
